primitives/arithmetic/src/fixed.rs

```diff
@@ -0,0 +1,1308 @@
+		self.into_inner() >= Self::Inner::zero()
+	}
+
+	/// Checks if the number is negative.
+	fn is_negative(self) -> bool {
+		self.into_inner() < Self::Inner::zero()
+	}
+
+	/// Returns the integer part.
+	fn trunc(self) -> Self;
+
+	/// Returns the fractional part.
+	fn frac(self) -> Self {
+		self.saturating_sub(self.trunc()).saturating_abs()
+	}
+
+	/// Returns the smallest integer greater than or equal to a number.
+	fn ceil(self) -> Self {
+		if self.is_negative() {
+			self.trunc()
+		} else {
+			self.saturating_add(Self::one()).trunc()
+		}
+	}
+
+	/// Returns the largest integer less than or equal to a number.
+	fn floor(self) -> Self {
+		if self.is_negative() {
+			self.saturating_sub(Self::one()).trunc()
+		} else {
+			self.trunc()
+		}
+	}
+
+	/// Returns the nearest integer to a number. Round half-way cases away from 0.0.
+	fn round(self) -> Self;
+}
+
+/// Data type used as intermediate step in some computations to avoid overflow.
+struct I129 {
+	value: u128,
+	negative: bool,
+}
+
+impl<N: FixedPointOperand> From<N> for I129 {
+	fn from(n: N) -> I129 {
+		if n < N::zero() {
+			let value: u128 = n.checked_neg()
+				.map(|n| n.unique_saturated_into())
+				.unwrap_or(N::max_value().unique_saturated_into().saturating_add(1));
+			I129 { value, negative: true }
+		} else {
+			I129 { value: n.unique_saturated_into(), negative: false }
+		}
+	}
+}
+
+/// Transforms an `I129` to `N` if it is possible.
+fn from_i129<N: FixedPointOperand>(n: I129) -> Option<N> {
+	let max_plus_one: u128 = N::max_value().unique_saturated_into().saturating_add(1);
+	if n.negative && N::min_value() < N::zero() && n.value == max_plus_one {
+		Some(N::min_value())
+	} else {
+		let unsigned_inner: N = n.value.try_into().ok()?;
+		let inner = if n.negative { unsigned_inner.checked_neg()? } else { unsigned_inner };
+		Some(inner)
+	}
+}
+
+/// Returns `R::max` if the sign of `n * m` is positive, `R::min` otherwise.
+fn to_bound<N: FixedPointOperand, D: FixedPointOperand, R: Bounded>(n: N, m: D) -> R {
+	if (n < N::zero()) != (m < D::zero()) {
+		R::min_value()
+	} else {
+		R::max_value()
+	}
+}
+
+macro_rules! implement_fixed {
+	(
+		$name:ident,
+		$test_mod:ident,
+		$inner_type:ty,
+		$div:tt,
+		$title:expr $(,)?
+	) => {
+		/// A fixed point number representation in the range [`Inner::min_value() / DIV`, `Inner::max_value() / DIV`].
+		///
+		#[doc = $title]
+		#[derive(Encode, Decode, Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+		pub struct $name($inner_type);
+
+		impl From<$inner_type> for $name {
+			fn from(int: $inner_type) -> Self {
+				$name::saturating_from_integer(int)
+			}
+		}
+
+		impl<N: FixedPointOperand, D: FixedPointOperand> From<(N, D)> for $name {
+			fn from(r: (N, D)) -> Self {
+				$name::saturating_from_rational(r.0, r.1)
+			}
+		}
+
+		impl FixedPointNumber for $name {
+			type Inner = $inner_type;
+
+			const DIV: Self::Inner = $div;
+
+			fn from_inner(inner: Self::Inner) -> Self {
```

Comment:
The documentation says

> /// Raw constructor. Assigns `int / Self::accuracy()` to `self`.

one is clearly wrong. If the document is wrong, then it is okay. If the other way around, then some test should catch this.

Reply:
Not sure I understand. If `int = 1` then `self = 1/1_000_000...` which I think is correct. Or is the wording? I thought it was more clear to write this way.

Reply:
yeah i read it differently, clear now. I got confused because in the implementation you name the parameter `int`. probably best to call it `inner`.